Concept robustness

Kama93 opened the pull request:
Dear exot,

In misc.clj, i`ve added the functionality to compute the robustness of concepts.
It is also possible to compute the average-concept-robustness of all concepts of a context.

I`ve also added a file socialanalytics.clj, which, right now, allows to:
-Compute the average-shortest-path of the following graphs for a given context:
  1.) The one which has as verticies the objects and the attributes and as edges the incidents.
   2.) The one which has as verticies the objects and in which two objects share an edge if they
       share an attribute.
   3.) The one which has as verticies the attributes and in which two attributes share an edge if they
       share an object.
-Compute the vertice-degrees of the same graphs as above.
For all those functions tests are implemented. I tried to stick to the style of the already existing tests.

Sincerly,
Kama93

exot commented:
Hi Kama93,

thanks for your work! :)  I will try to have a look at your work in the course of next week, but I would like to ask you to clean up the commit history a bit so that it's easier to understand (for me, at least ;) ):

- please do not include merge commits from dev, just rebase on dev instead
- could you try to squash some commits, or eliminate redundant commits (e.g., you had implemented a power-set function in one commit, only to remove it in a latter one; could you try to remove these two commits altogether?)

Please let me know whether you need any help with this!

Additionally, I would like to separate your implementation for socialanalysis from this pull request, to keep things clean.  Could you please move these commits to a separate branch and open a new pull request, please?  I currently vary whether applications for FCA such as Social Network Analysis should be included in conexp-clj proper, but then FCA lives from it's applications, doesn't it?  Maybe we should implement such applications in a new namespace at least, like conexp.fca.applications.  What do you say?

After these changes, I would like to merge your contributions to a feature branch first and make some further modifications.  I will let you know when I am done with this so that you can have a look before it's going to be included in dev.

Finally, don't forget to add yourself to the AUTHORS.md file! (however you want to, even if only with your pseudonym) :)

Thanks again for your efforts!

  Daniel

Kama93 commented:
Hi Daniel,
thanks for your answer. I`m going to squash and delete some (or probably many =) ) commits to make the history more understandable. As soon as I´m ready or if I need some help, I´ll get back to you.

To your point about conexp.fca.applications:
Do you want me to put a file called applications.clj in the folder FCA? Otherwise i could create another folder 'applications'  in FCA which has socialanlytics.clj as it first file?
I think that the file/folder conexp.fca.applications would have the potential to contain a lot of functionallity and grow a lot over time.
To keep things in a transparent structure I would therefore prefer the second option.

Kind Regards,
Max

exot commented:
Hey Max,

yes, I think my idea was more in the direction of your second proposal (although I have to admit I didn't spend much thought about it …).  So yes, please create a new folder `applications` under `conexp/fca` and add (in a separate pull request) your implementation of `socialanalytics.clj` in there.

Thanks! :)

  Daniel

Kama93 commented:
Hey Daniel,
I cleaned up the history of the robustness-branch. As it was my first time doing more than just push/pulll/commit with git it took me quite a long time. So sorry for the delay.
What I basically did:
- If I commite different changes in one commit I splitted this commit. That was neccessary as I worked on socialanalytics and concept-robustness sometimes for the same time.
- It was often the case that I created a function in a commit to rewrite it in a later commit to rewrite it again and finally fix the docstring in a commit. If that happened, I squashed them all together.
- Removed all commits belonging to socialanalytics as you asked me for.

Over the weekend I looked over my current implementation of socialanalytics.  And I was thinking about rewriting the functions and using a clojure graph-libary like loom. I think that would make the functions smaller and more understandable. And it would make it easier to add more functionality later. So I want to test if I could do this without loosing speed. I will do the other pull request with the socialanalytics implementation after this if that works for you.

Regards,
Max

Kama93 commented:
I`ve found some small mistakes in docstrings/comments. Also an assertion had a wrong string as it`s last parameter. I hope things are fine now...

Kama93 commented:
Hey Daniel,
I added some tests for eval-polynomial and pushed them to my concept-robustness branch.
Regards,
Max

exot commented:
Hi Max,

thanks for the tests, I have included them in corresponding feature branch.  I also made some cosmetic changes, and I would like to ask you to check whether I didn't break anything (all tests passed, though).

If you are fine, I will merge your changes to `dev`.

Best,

  Daniel

(PS: reopening this pull request because I closed it too early)

exot commented:
Good catch, thanks!

Merging into `dev` is done now, I am thus closing this pull request (this time for real! ;) )

Best,

  Daniel
